### PR TITLE
feat(plugins): inject git commit hash into bundles at build time

### PR DESCRIPTION
**Pull request by eskirk**
## Summary

- Adds \`resolveGitHash()\` to \`faro-bundlers-shared\` which resolves the commit hash at build time: explicit \`gitHash\` option → \`git rev-parse HEAD\` → \`bundleId\` if it is a valid 40-char hex SHA → \`undefined\`
- All three plugins (webpack, rollup, esbuild) inject \`window.__faroGitHash_<appName>\` alongside the existing \`window.__faroBundleId_<appName>\` when a hash resolves
- Adds \`faroGitHashSnippet()\` using the same safe-access IIFE pattern as \`faroBundleIdSnippet()\`
- Adds \`inject-git-hash\` CLI command mirroring \`inject-bundle-id\` for post-build injection workflows
- All plugins log via \`consoleInfoOrange\` when git hash resolution fails so the user knows the feature didn't activate

This is part of the suspected commits feature. Companion SDK PR: https://github.com/grafana/faro-web-sdk/pull/2039

## Test plan

- [ ] Build a webpack project with \`gitHash\` unset — confirm auto-detection via \`git rev-parse HEAD\` injects the correct hash
- [ ] Build with \`gitHash: process.env.GITHUB_SHA\` — confirm explicit value is used
- [ ] Build outside a git repo with no \`gitHash\` option — confirm \`[Faro] Git hash could not be resolved\` is logged and no \`__faroGitHash_\` variable is injected
- [ ] Build with \`bundleId\` set to a 40-char hex SHA and no \`gitHash\` — confirm bundleId value is used as the git hash
- [ ] Run \`faro-cli inject-git-hash -n myapp -f 'dist/**/*.js'\` and confirm the snippet is prepended

### CODE REVIEW

File: packages/faro-cli/src/index.ts [hunk 1]
```diff
@@ -793,5 +793,47 @@ export const injectBundleId = async (
...
+    try {
+      const stat = await fs.promises.stat(file);
+      if (!stat.isFile()) {
+        results.push({ file, modified: false, error: 'Not a regular file' });
+        continue;
+      }
+
+      const content = await fs.promises.readFile(file, 'utf8');
+
+      if (content.includes(`__faroGitHash_${appName}`)) {
```
Comment (kpelelis) [inline packages/faro-cli/src/index.ts]: maybe check if there is a value to it as well. Can it be that it looks like

`__faroGitHash_${appName}="";`

**Review by kpelelis — DISMISSED** [review]
Very nice work 🚀 🚀 . I am only requesting changes to make sure the one comment is addressed

**Review by codecapitano — COMMENTED** [review]
Multi-phase review — high-impact findings inline.

**Out-of-line note:** none of the npm READMEs (`packages/*/README.md`) document the new `gitHash` option or the `inject-git-hash` CLI command. npm/GitHub is the primary discovery surface; the website doc shouldn't be the sole reference. Worth a follow-up commit to mirror the website's options-table row in each plugin README + an "Injecting Git Hash" section in the CLI README.

Companion comments on faro-web-sdk#2039 and website#30561.

File: packages/faro-bundlers-shared/src/index.ts [hunk 2]
```diff
@@ -260,6 +262,33 @@ export const faroBundleIdSnippet = (bundleId: string, appName: string) => {
   return `(function(){try{var g=typeof window!=="undefined"?window:typeof global!=="undefined"?global:typeof self!=="undefined"?self:{};g["__faroBundleId_${appName}"]="${bundleId}"}catch(l){}})();`;
 };
 
+export const faroGitHashSnippet = (gitHash: string, appName: string) => {
+  return `(function(){try{var g=typeof window!=="undefined"?window:typeof global!=="undefined"?global:typeof self!=="undefined"?self:{};g["__faroGitHash_${appName}"]="${gitHash}"}catch(l){}})();`;
+};
```
Comment (codecapitano) [inline packages/faro-bundlers-shared/src/index.ts:320]: **Code injection sink in shipped bundles.** `gitHash` and `appName` are interpolated into a JS string literal / object key with no escaping. A value containing `"` or `\n` (e.g., `gitHash: '";alert(1);//'`) breaks out and executes in every page that loads the bundle. The `gitHash` value flows from `process.env.*` and user config — provenance is weaker than `bundleId`. The same shape exists in `faroBundleIdSnippet` today; worth hardening in a follow-up.

```suggestion
export const faroGitHashSnippet = (gitHash: string, appName: string) => {
  const key = JSON.stringify(`__faroGitHash_${appName}`);
  const value = JSON.stringify(gitHash);
  return `(function(){try{var g=typeof window!=="undefined"?window:typeof global!=="undefined"?global:typeof self!=="undefined"?self:{};g[${key}]=${value}}catch(l){}})();`;
};
```

File: packages/faro-bundlers-shared/src/index.ts [hunk 3]
```diff
@@ -260,6 +262,33 @@ export const faroBundleIdSnippet = (bundleId: string, appName: string) => {
...
+
+const GIT_SHA_PATTERN = /^[0-9a-f]{40}$/;
+
+export function resolveGitHash(options: { gitHash?: string; bundleId: string }): string | undefined {
+  if (options.gitHash) {
+    return options.gitHash;
```
Comment (codecapitano) [inline packages/faro-bundlers-shared/src/index.ts]: **Explicit `gitHash` is returned verbatim with no format check** while the auto-detect and bundleId-fallback paths are gated on `GIT_SHA_PATTERN`. A typo'd value (`gitHash: "main"`, `gitHash: "v1.2.3"`) ships into telemetry and downstream consumers can't rely on `app.gitHash` matching any contract.

Either validate here (`GIT_SHA_PATTERN.test(options.gitHash)` and reject/warn), or document explicitly that the explicit path is unvalidated. Pin the contract one way.

File: packages/faro-bundlers-shared/src/index.ts [hunk 4]
```diff
@@ -260,6 +262,33 @@ export const faroBundleIdSnippet = (bundleId: string, appName: string) => {
...
+  }
+
+  try {
+    const hash = execSync("git rev-parse HEAD", { encoding: "utf8", stdio: ["pipe", "pipe", "pipe"] }).trim();
```
Comment (codecapitano) [inline packages/faro-bundlers-shared/src/index.ts]: `execSync("git rev-parse HEAD")` has no `timeout` and inherits the parent process's `cwd`. A hung git (network FS, broken `core.hooksPath`, lock contention) freezes builds; running from an unrelated cwd silently injects HEAD of the wrong repo.

```suggestion
    const hash = execSync("git rev-parse HEAD", {
      encoding: "utf8",
      stdio: ["pipe", "pipe", "pipe"],
      timeout: 5000,
    }).trim();
```

Optional follow-up: accept a `cwd` option in `FaroSourceMapUploaderPluginOptions` for monorepo cases.

File: packages/faro-webpack/src/index.ts [hunk 5]
```diff
@@ -119,6 +122,10 @@ export default class FaroSourceMapUploaderPlugin
         ? options.maxUploadSize
         : THIRTY_MB_IN_BYTES;
     this.proxy = options.proxy;
+    this.gitHash = resolveGitHash({ gitHash: options.gitHash, bundleId: this.bundleId });
+    if (!this.gitHash) {
+      consoleInfoOrange(`Git hash could not be resolved. window.__faroGitHash_${this.appName} will not be injected.`);
```
Comment (codecapitano) [inline packages/faro-webpack/src/index.ts]: **Spec violation** + DX inconsistency. `git-hash-injection_spec.md` scenario "No git hash can be resolved" requires "no error or warning is emitted." `design.md` Decision 1 reiterates: "no warning, no error — to avoid noise in environments where git isn't available (e.g. Docker builds without `.git`)." This unconditional log fires on every Docker / no-`.git` CI build.

Same call exists in `faro-rollup/src/index.ts:48` and `faro-esbuild/src/index.ts:49`.

Two cleanest options:
1. Drop the log entirely (matches the spec).
2. Gate behind `verbose` (consistent with every other log here) and add a positive `verbose && consoleInfoOrange(\`Resolved gitHash <sha> via <source>\`)` next to `resolveGitHash` so users can debug both directions. The CLI already does this at `cli.ts:302`.

Either way, pick one and align spec + impl + docs.

File: packages/faro-cli/src/cli.ts [hunk 6]
```diff
@@ -267,4 +268,77 @@ Example:
     }
   });
 
+program
+  .command('inject-git-hash')
```
Comment (codecapitano) [inline packages/faro-cli/src/cli.ts:272]: **The CLI implements only steps 1 and 2 of the priority chain** (no `bundleId`-as-SHA fallback) and hard-exits when `git rev-parse HEAD` fails, while the bundler `resolveGitHash` falls through silently. A project building via the CLI in a `.git`-less Docker image fails outright; the same project via the bundler plugin succeeds.

Two paths to consistency:
1. Have the CLI call `resolveGitHash` (accepting a `--bundle-id` option) and exit cleanly when nothing resolves, mirroring the plugin contract per spec scenario "No git hash can be resolved".
2. If the CLI is intentionally stricter ("user explicitly invoked it, fail loudly"), document the divergence in the CLI README.

Either way, sharing `resolveGitHash` would prevent future drift between the two surfaces.

File: packages/faro-cli/src/cli.ts [hunk 7]
```diff
@@ -267,4 +268,77 @@ Example:
...
+        dryRun: options.dryRun,
+      });
+
+      const modifiedCount = results.filter((r) => r.modified).length;
+
+      if (options.dryRun) {
+        consoleInfoOrange(`Would modify ${modifiedCount} of ${matchedFiles.length} files`);
+      } else {
+        consoleInfoOrange(`Modified ${modifiedCount} of ${matchedFiles.length} files`);
+      }
```
Comment (codecapitano) [inline packages/faro-cli/src/cli.ts:333]: After the loop, `modifiedCount` is computed but per-file `error` strings collected at `index.ts:786-791, 831-834` are never inspected. If every file fails to write (EACCES, EROFS, read-only mount), the CLI prints "Modified 0 of N files" and exits 0 — CI green-lights a build with zero injection. Particularly bad because the website doc steers customers toward wiring this into CI.

```suggestion
      const modifiedCount = results.filter((r) => r.modified).length;
      const errorCount = results.filter((r) => r.error).length;

      if (options.dryRun) {
        consoleInfoOrange(`Would modify ${modifiedCount} of ${matchedFiles.length} files`);
      } else {
        consoleInfoOrange(`Modified ${modifiedCount} of ${matchedFiles.length} files`);
      }

      if (errorCount > 0) {
        console.error(`Error: ${errorCount} file(s) failed to inject git hash`);
        process.exit(1);
      }
```

(Same pattern is worth applying to `inject-bundle-id` in a follow-up.)

File: packages/faro-webpack/src/test/index.test.ts [hunk 8]
```diff
@@ -184,6 +184,33 @@ describe("Faro Webpack Plugin", () => {
     expect(firstCharsPos).toBeLessThan(200);
   });
 
+  test("gitHash snippet is injected when gitHash option is provided", async () => {
```
Comment (codecapitano) [inline packages/faro-webpack/src/test/index.test.ts:187]: A few headline scenarios are missing test coverage:

1. **bundleId-as-SHA fallback** — no test passes a 40-char hex `bundleId` without `gitHash` to verify `resolveGitHash`'s third branch. The branch is currently dead-code-equivalent: it could be deleted without breaking any test.
2. **Auto-detect happy path** — `child_process.execSync` is globally mocked to throw at the top of this file (and identically in rollup/esbuild test files). No test re-mocks it to return a fake SHA and asserts injection. The "explicit beats auto-detect" precedence is therefore also untested.
3. **Boundary cases for the regex** — 39-char hex, 41-char, uppercase hex, auto-generated bundleId. Loosening `GIT_SHA_PATTERN` would not be caught.
4. **No tests for the new `inject-git-hash` CLI command** — option parsing, glob matching, file rewriting, dry-run, exit semantics all unverified. Mirror the existing `inject-bundle-id` test shape and add the all-files-fail case from the comment on `cli.ts:331`.

Same gaps in `faro-rollup/src/test/index.test.ts` and `faro-esbuild/src/test/index.test.ts`.

File: packages/faro-cli/src/cli.ts (hunk 7 above)
Comment (eskirk) [reply packages/faro-cli/src/cli.ts:333]: I dont know if failing to inject the git hash should exit the process with a failure - maybe I need to add a "exitOnFailure" property or something like that?

File: packages/faro-cli/src/cli.ts (hunk 6 above)
Comment (eskirk) [reply packages/faro-cli/src/cli.ts:272]: I have removed the bundle ID method for injecting the git hash, I realized it was not necessary. I did make the CLI share the `resolveGitHash` helper - that should be helpful.

File: packages/faro-cli/src/cli.ts (hunk 7 above)
Comment (eskirk) [reply packages/faro-cli/src/cli.ts:333]: ok actually, on second thought - it should exit with a failure. that is the entire point of running that command 😅

**Review by codecapitano — DISMISSED** [review]

**Review by martinkuba — APPROVED** [review]
